### PR TITLE
Propose traced transaction construction

### PR DESCRIPTION
**Pull request by ryoqun**
(boring meta-joke due to imminent sleep time.)

#### Problem

This is half-done!

#### Summary of Changes

Push more commitz and squaz

context (original idea is written when the itoken reviewing: https://github.com/solana-labs/solana/pull/15927#discussion_r595842064)

### CODE REVIEW

File: docs/src/proposals/traced-transaction-construction.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
+# Traced transaction construction via RPC's `simulateTransaction`
+
+## Problem
+
+With more composition comes in the Solana's program ecosystem, it will be harder
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:5]: this doc is still in the draft phase. but I think the first half of this is written decently. ;)

File: docs/src/proposals/traced-transaction-construction.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+breaks compatibility?
+
+### Run these before replaying stage on validator
+
+unavoidable latency.  (and scheduling difficulty).
+The proposed implementation doesn't have such a downside
+
+## Future possible expansions
+
+- Wallet account balance delta check: append VerifyTokenBalances ix at last and pass the expected balances of spl-tokens or hash of it (also with delegation status).
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:246]: this is my (hopefully nice) solution to this: https://github.com/solana-labs/solana/issues/17762

File: docs/src/proposals/traced-transaction-construction.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+- Browser-side bpf program simualtion for server-less architecture with online wasm translation
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:247]: this is only possible because solana tx execution doesn't touch the whole account state in the chain. :)

File: docs/src/proposals/traced-transaction-construction.md (hunk 2 above)
Comment (ryoqun) [reply docs/src/proposals/traced-transaction-construction.md:246]: also, dynamic ix append as directed traced simulation might also be useful.

File: docs/src/proposals/traced-transaction-construction.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
++    let associated_token_account_info = access_as_unsigned(associated_token_address, READ_WRITE)?;
+ 
+     let associated_token_account_signer_seeds: &[&[_]] = &[
+         &wallet_account_info.key.to_bytes(),
+```
+
+This reduces the number of required accounts for CPI and clients from 7 to 3, reducing by 4.
+(Obviously, we can't actually deploy these changes because this will break
+compatibility. This is only for illustration of effective applicability to
+moderately-complex program construct.)
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:122]: also, maybe worth to mention that this doesn't reduce the resultant tx binary size, we need still this badly for more composability: https://github.com/solana-labs/solana/pull/15927 cc: @jstarry 

The good thing is that this nor the big tx block each other. :)

File: docs/src/proposals/traced-transaction-construction.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+race condition around program deploy (= possible account loading change) => just retry also seek to the ix versioning for perfection
+
+(traced program just load dummy addresses for tx versionsing for completeness)
+
+## careful consideration
+
+prevent program  detect to be simulated or really-executed
+
+
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:232]: also, write about the strip of is_signer and no use case limited by this stripping.

File: docs/src/proposals/traced-transaction-construction.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+construction for CPI) or off-chain (transaction construction by end-user
+interfaces) boundaries likewise.
+
+Lastly, it's practically impossible to change those accessed account addresses
+once deployed while retaining compatibility, given the fact the consumers are
+expected to hard-code the account addresses for consumption, even if the
+deployed programs can be now upgradeable. This severely hampers ongoing
+consumed-program's implementation refinement in the areas of data migration and
+improved concurrency, by preventing transparent account switching and
+sharding with vectorized accounts, respectively.
```
Comment (jstarry) [inline docs/src/proposals/traced-transaction-construction.md:26]: what is meant by "vectorized accounts"?

Comment (ryoqun) [reply docs/src/proposals/traced-transaction-construction.md:26]: @jstarry don't be bothered with it.. ;) well, I used this fancy adjective, remotely referencing the usage like `vectorization with SSE4/AVX2`. Maybe `sharding` by itself should be enough.

`Sharding` here means avoiding write lock to the same internal account by using different accounts to write by deriving from the first byte of other account addresses like order_id for DEX. So, program can avoid write lock contention easily. Also, see the sharding example for code.

I want to say, hiding internal account addresses in this way makes these optimizations can be introduced by drop-in basis.

Comment (jstarry) [reply docs/src/proposals/traced-transaction-construction.md:26]: Ah got it, I like the idea of abstracting away implementation details like this, thanks for the explanation!

File: docs/src/proposals/traced-transaction-construction.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+in no-op and harmless if all of the executed programs aren't aware of this new
+functionality.
+
+## Traced syscalls
+
+The runtime exposes new syscall to provide new account access mechanism which
+can be traced:
+
+```rust
+fn access_as_unsigned(address: &Pubkey, flags = enum { READ, WRITE}):
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:53]: Considering this https://github.com/solana-labs/solana/pull/17901,

Maybe we should augment this by returning Enum<AccountInfo, AccountRef> with new flag ACCESS_AS_REF. so that we can avoid any serializing into program execution environment.

AccountRef can only be used for passing to child cpi.

File: docs/src/proposals/traced-transaction-construction.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+) -> ProgramResult {
+    let counter_address = Pubkey::find_program_address(
+        &[  
+            &program_id.to_bytes(),
+            "counter",
+        ],
+        program_id,
+    );
+   let transfer_counter = access_as_unsigned(counter_address, READ_WRITE)?;
+   let clock_sysvar = access_as_unsigned(sysvars::clock::id(), READ_WRITE)?;
```
Comment (jackcmay) [inline docs/src/proposals/traced-transaction-construction.md:141]: Why is clock rw?

File: docs/src/proposals/traced-transaction-construction.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    expected_decimals: Option<u8>,    
+) -> ProgramResult {
+    let counter_address = Pubkey::find_program_address(
+        &[  
+            &program_id.to_bytes(),
+            "counter",
+        ],
+        program_id,
+    );
+   let transfer_counter = access_as_unsigned(counter_address, READ_WRITE)?;
```
Comment (jackcmay) [inline docs/src/proposals/traced-transaction-construction.md:140]: What type is `transfer_counter`?

**Comment by jackcmay** [comment]
This is interesting.  I like that it reduces and simplifies a program's calling interface by hiding internal-only accounts that the caller doesn't need to be or shouldn't be aware of.

**Comment by jackcmay** [comment]
In addition to a syscall used to discover dependent accounts, if any cpi is invoked during simulation that specifies new accounts, those new accounts should also be added to the list.  During non-simulation those dependent accounts would be listed in the tx, loaded along with the rest of the accounts and referenced as needed.

File: docs/src/proposals/traced-transaction-construction.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+```patch
+diff --git a/associated-token-account/program/src/processor.rs b/associated-token-account/program/src/processor.rs
+index 98eb08b..fa8ae50 100644
+--- a/associated-token-account/program/src/processor.rs
++++ b/associated-token-account/program/src/processor.rs
+@@ -22,13 +22,12 @@ pub fn process_instruction(
+     let account_info_iter = &mut accounts.iter();
+ 
+     let funder_info = next_account_info(account_info_iter)?;
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:91]: related to #18588

File: docs/src/proposals/traced-transaction-construction.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
+# Traced transaction construction via RPC's `simulateTransaction`
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:1]: traced => instrumented?

**Comment by ruuda** [comment]
Tracing can only reveal references in code that is actually executed. Consider something like

```rust
let clock = Clock::get()?;
if clock.slot % 2 == 0 {
    reference_account_a();
} else {
    reference_account_b();
}
```

If the RPC traces the program execution and says it referenced account A, then it is likely that by the time it executes for real, the program will try to reference account B.

This is similar to how tests can only reveal bugs in code that is executed; to get an exhaustive answer, you need some kind of (static) program analysis or abstract interpretation.

**Comment by ryoqun** [comment]
> Tracing can only reveal references in code that is actually executed. Consider something like
> 
> ```rust
> let clock = Clock::get()?;
> if clock.slot % 2 == 0 {
>     reference_account_a();
> } else {
>     reference_account_b();
> }
> ```
> 
> If the RPC traces the program execution and says it referenced account A, then it is likely that by the time it executes for real, the program will try to reference account B.
> 
> This is similar to how tests can only reveal bugs in code that is executed; to get an exhaustive answer, you need some kind of (static) program analysis or abstract interpretation.

@ruuda thanks for chiming in! you're first penguin from community. I really appreciate that.

Yeah, I understand the limitation. You have a point. But I still think this proposal could be reasonable solution for the problems described, at the cost of the strict requirement of rpc's simulation when issuing transactions to any advanced programs.

Maybe I should have added a prose like this?:

```
# Limitations (or Compromises)

This facility must not be abused and used wisely, needing proper developer
guidelines/communication. Firstly, dynamic account selection logic shouldn't
be timing sensitive or not change frequently. Because there is unavoidable
timing delays between simulation and actual execution on chain, transactions
could fail unexpectedly if implemented badly.

In other words, while dynamic account loading could give greater flexibility, the
logic should rather be static over time and for individual user sessions. The dynamic loading
isn't goal par itself. Rather this proposal tries to reduce and simplify a program's
calling interface by hiding internal-only accounts that the caller doesn't need to be or
shouldn't be aware of.

# Alternatives

- Some (static) IDL/DSL for account loading logic: hard to develop/design to cover any conceivable use cases. extra complexity
- provide no programmatic/systemic way to describe account loading (status quo): hard to compose defi programs. 
```

How do you like?

File: docs/src/proposals/traced-transaction-construction.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+
+
+
+also, page-indexed account can be incorporated to find compact combination of existing account indexes to match the required account list
+
+## downside
+
+race condition around program deploy (= possible account loading change) => just retry also seek to the ix versioning for perfection
+
+(traced program just load dummy addresses for tx versionsing for completeness)
```
Comment (ryoqun) [inline docs/src/proposals/traced-transaction-construction.md:226]: also, I have to embed this pr comment into doc: https://github.com/solana-labs/solana/pull/17796#issuecomment-885060430

**Comment by ruuda** [comment]
> How do you like?

I don’t have a strong opinion on it, I just wanted to point out the limitation. Personally I think that with the way the Solana runtime is exposed to Rust at the moment, Rust is not a great language for writing Solana programs, and in the long run I expect some DSL to take over. A DSL could take care of those details internally, with a compiler figuring out which accounts are referenced.
